ssl/quic/quic_wire.c

```diff
@@ -366,18 +367,45 @@ int ossl_quic_wire_encode_frame_path_response(WPACKET *pkt,
 int ossl_quic_wire_encode_frame_conn_close(WPACKET *pkt,
                                            const OSSL_QUIC_FRAME_CONN_CLOSE *f)
 {
-    if (!encode_frame_hdr(pkt, f->is_app ? OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_APP
-                                         : OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_TRANSPORT)
-            || !WPACKET_quic_write_vlint(pkt, f->error_code))
+    const unsigned int is_1d = f->is_app;
+
+    /*
+     * RFC 9000 10.2.3: A CONNECTION_CLOSE of type 0x1d MUST be replaced
+     * by a CONNECTION_CLOSE of type 0x1c when sending the frame in Initial
+     * or Handshake packets.
+     */
+    if (!encode_frame_hdr(pkt, is_1d ? OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_APP
+                                     : OSSL_QUIC_FRAME_TYPE_CONN_CLOSE_TRANSPORT))
         return 0;
 
-    if (!f->is_app && !WPACKET_quic_write_vlint(pkt, f->frame_type))
+    /*
+     * RFC 9000 10.2.3: Endpoints ... SHOULD use the APPLICATION_ERROR
+     * code when converting to a CONNECTION_CLOSE of type 0x1c.
+     */
+     if (!WPACKET_quic_write_vlint(pkt, is_1d ? f->error_code
+                                              : QUIC_ERR_APPLICATION_ERROR))
         return 0;
 
-    if (!WPACKET_quic_write_vlint(pkt, f->reason_len)
-            || !WPACKET_memcpy(pkt, f->reason, f->reason_len))
+    /*
+     * RFC 9000 19.19: The application-specific variant of CONNECTION_CLOSE
+     * (type 0x1d) does not include this field.
+     */
+    if (!is_1d && !WPACKET_quic_write_vlint(pkt, f->frame_type))
         return 0;
 
+    /*
+     * RFC 9000 10.2.3: Endpoints MUST clear the value of the Reason
+     * Phrase field ... when converting to a CONNECTION_CLOSE of type 0x1c.
+     */
+    if (!is_1d) {
+        if (!WPACKET_quic_write_vlint(pkt, 0))
+            return 0;
```

Comment:
This is wrong and should not be done on this level. See https://github.com/openssl/openssl/pull/21135/commits/741a3abd92e120bb57f2f3bbc89bc395c5149829